modules/caddyhttp/reverseproxy/selectionpolicies.go

```diff
@@ -397,13 +400,12 @@ func leastRequests(upstreams []*Upstream) *Upstream {
 		return nil
 	}
 	var best []*Upstream
-	var bestReqs int
+	var bestReqs int = -1
 	for _, upstream := range upstreams {
 		reqs := upstream.NumRequests()
-		if reqs == 0 {
-			return upstream
-		}
-		if reqs <= bestReqs {
+		// If bestReqs was just initialized to -1
+		// we need to append upstream also
+		if reqs <= bestReqs || bestReqs == -1 {
```

Comment:
I don't really understand this; if an upstream has 0 requests, it is already the least it can be, therefore we can use it.

Reply:
Actually, there is 2 problems:

- Firstly, with the old implementation of function leastRequests, there is an issue when all upstream server numRequest is at least 1: as bestReqs is initialized to 0, this condition (`reqs <= bestReqs`) can never be true, so the function return nil and we have a 502 http response code
(see [here](https://github.com/caddyserver/caddy/blob/master/modules/caddyhttp/reverseproxy/selectionpolicies.go#L187) where is the right condition for the same problem)

It is easily reproducible with the unit test I add (TestLeastRequests) or by doing some concurrent http request.




- Secondly, -by replacing `choices[j] = upstream` by `choices = append(choices, upstream)`, I observed that there were no more random, and it was always the first upsteam server that was chosen when the numRequest was 0.
Removing the `if reqs == 0 { return upsteam }` was a way to solve that.

Now, I realize that using append was a bad idea: the pool with random host will always contains the first upstreams and the pool size will not always match the configured size.


So I will rollback this change, and instead I'm gonna bypass nil upstream on leastRequests function